In [172]:
import pandas as pd
import os

In [174]:
df_main = pd.read_csv("senegal_weather_weekly_2010_2019.csv")

If 52 weeks, then


1 month	4.345 weeks

2 months	8.69 weeks

3 months	13.04 weeks

4 months	17.38 weeks

5 months	21.73 weeks

6 months	26.07 weeks

7 months	30.42 weeks

8 months	34.76 weeks

9 months	39.11 weeks

10 months	43.45 weeks

11 months	47.8 weeks

12 months	52.14 weeks

If 53 weeks, then

1 month	4.416 weeks

2 months	8.833 weeks

3 months	13.25 weeks

4 months	17.667 weeks

5 months	22.083 weeks

6 months	26.5 weeks

7 months	30.917 weeks

8 months	35.333 weeks

9 months	39.75 weeks

10 months	44.167 weeks

11 months	48.583 weeks

12 months	53 weeks

In [175]:
w2m_52 = {"0":0, "1":4, "2":8, "3":13, "4":17, "5":21, "6":26, "7":30, "8":34, "9":39, "10":43, "11":47, "12":52}

w2m_53 = {"0":0, "1":4, "2":8, "3":13, "4":17, "5":22, "6":26, "7":30, "8":35, "9":39, "10":44, "11":48, "12":53}

years_52 = [2010, 2011, 2012, 2013, 2014, 2016, 2017, 2018, 2019]
years_53 = [2015]

In [176]:
df_2010 = df_main[df_main["year"]==2010]
df_2011 = df_main[df_main["year"]==2011]
df_2012 = df_main[df_main["year"]==2012]
df_2013 = df_main[df_main["year"]==2013]
df_2014 = df_main[df_main["year"]==2014]
df_2015 = df_main[df_main["year"]==2015]
df_2016 = df_main[df_main["year"]==2016]
df_2017 = df_main[df_main["year"]==2017]
df_2018 = df_main[df_main["year"]==2018]
df_2019 = df_main[df_main["year"]==2019]

In [177]:
df_2015

year  week     gpslat     gpslon       temp   rel_hum        hum  \
260     2015     1  13.988077 -15.839605  22.460000  0.297143  21.350000   
261     2015     2  13.988077 -15.839605  23.722857  0.324286  23.492857   
262     2015     3  13.988077 -15.839605  24.277143  0.332857  24.322857   
263     2015     4  13.988077 -15.839605  24.597143  0.308571  24.304286   
264     2015     5  13.988077 -15.839605  24.378571  0.261429  23.240000   
...      ...   ...        ...        ...        ...       ...        ...   
424402  2015    49  13.826203 -16.133960  24.767143  0.418571  26.498571   
424403  2015    50  13.826203 -16.133960  24.315714  0.430000  26.038571   
424404  2015    51  13.826203 -16.133960  24.277143  0.374286  25.014286   
424405  2015    52  13.826203 -16.133960  24.247143  0.385714  25.171429   
424406  2015    53  13.826203 -16.133960  24.107143  0.368571  24.725714   

        wind_speed   wind_dir  solar_radiat  thermal_radiat  dir_solar_radiat  \
260       3.091429  51.417143    212.285714      329.571429        134.861429   
261       2.570000  28.418571    211.285714      341.714286        138.201429   
262       2.624286  18.790000    223.285714      338.857143        149.528571   
263       3.021429  26.317143    227.000000      348.714286        158.532857   
264       3.307143  79.282857    236.428571      330.714286        161.862857   
...            ...        ...           ...             ...               ...   
424402    2.388571  52.931429    232.714286      340.714286        178.444286   
424403    2.128571  32.131429    202.571429      352.000000        125.895714   
424404    2.610000  49.522857    223.571429      337.000000        163.158571   
424405    2.811429  89.671429    189.285714      358.142857        114.662857   
424406    2.915714  52.311429    186.285714      359.714286        108.451429   

           soil_1     soil_2     soil_3     soil_4   pr  
260     22.420000  23.040000  25.024286  28.028571  0.0  
261     23.744286  23.844286  24.851429  27.651429  0.0  
262     23.847143  23.905714  24.811429  27.264286  0.0  
263     24.365714  24.481429  24.920000  26.931429  0.0  
264     24.231429  24.182857  25.035714  26.674286  0.0  
...           ...        ...        ...        ...  ...  
424402  24.698571  25.225714  26.565714  28.264286  0.0  
424403  24.297143  24.870000  26.060000  27.997143  0.0  
424404  23.898571  24.368571  25.512857  27.692857  0.0  
424405  24.024286  24.537143  25.282857  27.374286  0.0  
424406  24.021429  24.360000  25.177143  27.090000  0.0  

[43195 rows x 17 columns]

In [181]:
def create_new_df(df_func, weeks, weeks_to_month, year):
    row_list=[]
    df_func = df_func[df_func["year"]==year]
    num_locations = int(len(df_func)/weeks)
    print(num_locations)  
    for loc in range(num_locations):
        loc_df = df_func[loc*weeks:(loc+1)*weeks]
        #print(len(loc_df))
        #lat, long = loc_df["gpslat"][0], loc_df["gpslon"][0]
        
        for i in range(12):#12): #i-i+1
            avg_df = loc_df[weeks_to_month[str(i)]:weeks_to_month[str(i+1)]].mean()

            row_list.append({"year":year, "month":(i+1),"gpslat":avg_df["gpslat"], "gpslon":avg_df["gpslon"], "temp":avg_df["temp"],
                          "rel_hum":avg_df["rel_hum"], "hum":avg_df["hum"], "wind_speed":avg_df["wind_speed"], 
                           "wind_dir":avg_df["wind_dir"], "solar_radiat":avg_df["solar_radiat"],
                           "thermal_radiat":avg_df["thermal_radiat"], "dir_solar_radiat":avg_df["dir_solar_radiat"],
                          "soil_1":avg_df["soil_1"], "soil_2":avg_df["soil_2"], "soil_3":avg_df["soil_3"],
                           "soil_4":avg_df["soil_4"], "pr":avg_df["pr"]})
            
    return row_list
            

year                2010.000000

week                   2.500000

gpslat                13.988077

gpslon               -15.839605

temp                  24.978571

rel_hum                0.318929

hum                   25.049286

wind_speed             2.577857

wind_dir              79.516071

solar_radiat         226.000000

thermal_radiat       336.500000

dir_solar_radiat     154.652500

soil_1                24.552143

soil_2                24.639286

soil_3                25.301786

soil_4                26.843214

pr                     0.000357

In [182]:
#new_df = pd.DataFrame(columns=[["year"], ["month"], ["gpslat"], ["gpslon"], ["temp"],
#                                   ["rel_hum"], ["hum"], ["wind_speed"], ["wind_dir"], ["solar_radiat"],
#                                   ["thermal_radiat"], ["dir_solar_radiat"], ["soil_1"], ["soil_2"], ["soil_3"], ["soil_4"], ["pr"]])


#new_df = pd.DataFrame(create_new_df(df_2010, 52, w2m, 2010))

In [183]:
for year in years_52:
    df_new = pd.DataFrame(create_new_df(df_main, 52, w2m_52, year))
    df_new.to_csv('monthly_senegal_data.csv', mode='a', header=False)


for year in years_53:
    df_new = pd.DataFrame(create_new_df(df_main, 53, w2m_53, year))
    df_new.to_csv('monthly_senegal_data.csv', mode='a', header=False)

815
815
815
815
815
815
815
815
815
815


In [131]:
df_new

[{'year': 2010,
  'month': 1,
  'gpslat': 13.98807667,
  'gpslon': -15.839605,
  'temp': 24.97857142857143,
  'rel_hum': 0.31892857142857145,
  'hum': 25.04928571428571,
  'wind_speed': 2.5778571428571433,
  'wind_dir': 79.51607142857142,
  'solar_radiat': 226.0,
  'thermal_radiat': 336.5,
  'dir_solar_radiat': 154.65249999999997,
  'soil_1': 24.552142857142854,
  'soil_2': 24.63928571428572,
  'soil_3': 25.301785714285714,
  'soil_4': 26.84321428571429,
  'pr': 0.00035714285714285714},
 {'year': 2010,
  'month': 2,
  'gpslat': 13.98807667,
  'gpslon': -15.839605,
  'temp': 27.547857142857143,
  'rel_hum': 0.40750000000000003,
  'hum': 30.235357142857147,
  'wind_speed': 2.5725000000000002,
  'wind_dir': 153.4925,
  'solar_radiat': 259.39285714285717,
  'thermal_radiat': 355.17857142857144,
  'dir_solar_radiat': 192.10464285714286,
  'soil_1': 27.712857142857143,
  'soil_2': 27.08642857142857,
  'soil_3': 26.48071428571429,
  'soil_4': 26.385714285714286,
  'pr': 0.0},
 {'year': 2010,
